test/common/io/io_uring_worker_impl_integration_test.cc

```diff
@@ -0,0 +1,270 @@
+    EXPECT_EQ(inet_pton(AF_INET, "127.0.0.1", &listen_addr.sin_addr.s_addr), 1);
+    EXPECT_EQ(Api::OsSysCallsSingleton::get()
+                  .bind(listen_socket_, reinterpret_cast<struct sockaddr*>(&listen_addr),
+                        sizeof(listen_addr))
+                  .return_value_,
+              0);
+    EXPECT_EQ(Api::OsSysCallsSingleton::get().listen(listen_socket_, 5).return_value_, 0);
+  }
+  void connect() {
+    struct sockaddr_in listen_addr = getListenSocketAddress();
+    Api::OsSysCallsSingleton::get().connect(
+        client_socket_, reinterpret_cast<struct sockaddr*>(&listen_addr), sizeof(listen_addr));
+    EXPECT_EQ(errno, EINPROGRESS);
+  }
+  void accept() {
+    auto file_event = dispatcher_->createFileEvent(
+        listen_socket_,
+        [this](uint32_t events) {
+          EXPECT_EQ(events, Event::FileReadyType::Read);
+          struct sockaddr_in server_addr;
+          socklen_t server_addr_len = sizeof(server_addr);
+          server_socket_ =
+              Api::OsSysCallsSingleton::get()
+                  .accept(listen_socket_, reinterpret_cast<struct sockaddr*>(&server_addr),
+                          &server_addr_len)
+                  .return_value_;
+          EXPECT_TRUE(SOCKET_VALID(server_socket_));
+          // The server socket should block.
+          int flags = fcntl(server_socket_, F_GETFL);
+          fcntl(server_socket_, F_SETFL, flags & ~O_NONBLOCK);
+        },
+        Event::PlatformDefaultTriggerType, Event::FileReadyType::Read);
+    while (!SOCKET_VALID(server_socket_)) {
+      dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+    }
+    file_event.reset();
+  }
+
+  void runToClose(os_fd_t fd) {
+    do {
+      dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+    } while (fcntl(fd, F_GETFD) == 0);
+  }
+
+  bool should_skip_{false};
+  os_fd_t listen_socket_{INVALID_SOCKET};
+  os_fd_t server_socket_{INVALID_SOCKET};
+  os_fd_t client_socket_{INVALID_SOCKET};
+  Api::ApiPtr api_;
+  Event::DispatcherPtr dispatcher_;
+  Event::GlobalTimeSystem time_system_;
+  std::unique_ptr<IoUringWorkerTestImpl> io_uring_worker_;
+};
+
+TEST_F(IoUringWorkerIntegrationTest, Injection) {
+  initialize();
+  initializeSockets();
+
+  auto& socket =
+      dynamic_cast<IoUringSocketTestImpl&>(io_uring_worker_->addTestSocket(server_socket_));
+  EXPECT_EQ(io_uring_worker_->getSockets().size(), 1);
+
+  socket.injectCompletion(Request::RequestType::Accept);
+  socket.injectCompletion(Request::RequestType::Cancel);
+  socket.injectCompletion(Request::RequestType::Close);
+  socket.injectCompletion(Request::RequestType::Connect);
+  socket.injectCompletion(Request::RequestType::Read);
+  socket.injectCompletion(Request::RequestType::Write);
+  socket.injectCompletion(Request::RequestType::Shutdown);
+
+  // Wait for server socket receive injected completion.
+  while (socket.accept_result_ == -1 || socket.cancel_result_ == -1 || socket.close_result_ == -1 ||
```

Comment:
seems like the above injects 7 ops, but we only check for 6 here?

Reply:
emm, good catch, indeed missing one, thanks!